ethdb/privateapi/ethbackend.go

```diff
@@ -271,70 +279,157 @@ func (s *EthBackendServer) EngineExecutePayloadV1(ctx context.Context, req *type
 	}
+
 	// Check if parent equate to the head
 	parent := gointerfaces.ConvertH256ToHash(req.Forkchoice.HeadBlockHash)
 	tx, err := s.db.BeginRo(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	headHeader, err := rawdb.ReadHeaderByHash(tx, parent)
-	if err != nil {
-		return nil, err
+	if parent != rawdb.ReadHeadHeaderHash(tx) {
+		// TODO(enriavil1): make unwind happen
+		return &remote.EngineForkChoiceUpdatedReply{
+			Status: string(Syncing),
+		}, nil
 	}
 
-	if atomic.LoadUint32(s.waitingForBeaconChain) == 0 || headHeader == nil {
+	// Same if we are not waiting for the beacon chain
+	if atomic.LoadUint32(s.waitingForBeaconChain) == 0 {
 		return &remote.EngineForkChoiceUpdatedReply{
 			Status: string(Syncing),
 		}, nil
 	}
 
-	// Hash is incorrect because mining archittecture has yet to be implemented
+	if !s.proposing {
+		return nil, fmt.Errorf("execution layer not running as a proposer. enable --proposer flag on startup")
+	}
+
 	s.pendingPayloads[s.payloadId] = types2.ExecutionPayload{
-		ParentHash:    req.Forkchoice.HeadBlockHash,
-		Coinbase:      req.Prepare.FeeRecipient,
-		Timestamp:     req.Prepare.Timestamp,
-		Random:        req.Prepare.Random,
-		StateRoot:     gointerfaces.ConvertHashToH256(headHeader.Root),
-		ReceiptRoot:   gointerfaces.ConvertHashToH256(types.EmptyRootHash),
-		LogsBloom:     &types2.H2048{},
-		GasLimit:      headHeader.GasLimit,
-		GasUsed:       0,
-		BlockNumber:   headHeader.Number.Uint64() + 1,
-		ExtraData:     []byte{},
-		BaseFeePerGas: &types2.H256{},
-		BlockHash:     gointerfaces.ConvertHashToH256(headHeader.Hash()),
-		Transactions:  [][]byte{},
+		Random:    req.Prepare.Random,
+		Timestamp: req.Prepare.Timestamp,
+		Coinbase:  req.Prepare.FeeRecipient,
 	}
-	// successfully assembled the payload and assinged the correct id
+	s.condPauseAssemble.Broadcast()
+	// successfully assembled the payload and assigned the correct id
 	defer func() { s.payloadId++ }()
 	return &remote.EngineForkChoiceUpdatedReply{
 		Status:    "SUCCESS",
 		PayloadId: s.payloadId,
 	}, nil
 }
 
+func (s *EthBackendServer) StartProposer() {
+
+	go func() {
+		s.condPauseAssemble.L.Lock()
+		defer s.condPauseAssemble.L.Unlock()
+
+		for {
+			// Wait until we need to process new payloads
+			s.condPauseAssemble.Wait()
+			s.mu.Lock()
+			// Go over each payload and re-update them
+			for id := range s.pendingPayloads {
+				// If we already assembled this block, let's just skip it
+				if s.pendingPayloads[id].BlockNumber != 0 {
+					continue
+				}
+				// we do not want to make a copy of the payload in the loop because it contains a lock
+				random := gointerfaces.ConvertH256ToHash(s.pendingPayloads[id].Random)
+				coinbase := gointerfaces.ConvertH160toAddress(s.pendingPayloads[id].Coinbase)
+				timestamp := s.pendingPayloads[id].Timestamp
+				// Tell the stage headers to leave space for the write transaction for mining stages
+				s.skipCycleHack <- struct{}{}
```

Comment:
`assemblePayloadPOS()` can take a long time. It's better to release the lock before the invocation and then re-acquire the lock.